### PR TITLE
[SPARK-11179] [SQL] Push filters through aggregate

### PR DESCRIPTION
**Pull request by nitin2goyal**
Push conjunctive predicates though Aggregate operators when their references are a subset of the groupingExpressions.

Query plan before optimisation :-
Filter ((c#138L = 2) && (a#0 = 3))
 Aggregate [a#0], [a#0,count(b#1) AS c#138L]
  Project [a#0,b#1]
   LocalRelation [a#0,b#1,c#2]

Query plan after optimisation :-
Filter (c#138L = 2)
 Aggregate [a#0], [a#0,count(b#1) AS c#138L]
  Filter (a#0 = 3)
   Project [a#0,b#1]
    LocalRelation [a#0,b#1,c#2]

### CODE REVIEW

**Comment by marmbrus** [comment]
Tests please.  Look [here](https://github.com/davies/spark/tree/byte_buffer/sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer) for examples.

**Comment by nitin2goyal** [comment]
Added tests

**Comment by hvanhovell** [comment]
We could do a similar thing for window functions.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala [hunk 1]
```diff
@@ -675,6 +676,20 @@ object PushPredicateThroughGenerate extends Rule[LogicalPlan] with PredicateHelp
...
 /**
+ * Push [[Filter]] operators through [[Aggregate]] operators in case predicate's
+ * refrence attributes are subset of groupBy attribute set.
+ */
+object PushPredicateThroughAggregate extends Rule[LogicalPlan] with PredicateHelper {
+
+  def apply(plan: LogicalPlan): LogicalPlan = plan transform {
+    case filter @ Filter(condition,
+    aggregate @ Aggregate(groupingExpressions, aggregateExpressions, child))
+        if (filter.condition.references -- AttributeSet(groupingExpressions)).isEmpty =>
```
Comment (rxin) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala]: rather than checking for complete overlap, can we pull out the expressions for group by columns and push those down?

Comment (nitin2goyal) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala]: Good point. Incorporated.

**Comment by marmbrus** [comment]
ok to test

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala [hunk 2]
```diff
@@ -675,6 +676,29 @@ object PushPredicateThroughGenerate extends Rule[LogicalPlan] with PredicateHelp
...
 /**
+ * Push [[Filter]] operators through [[Aggregate]] operators. Parts of the predicate that reference
+ * attributes which are subset of group by attribute set of [[Aggregate]] will be pushed beneath, 
+ * and the rest should remain above.
+ */
+object PushPredicateThroughAggregate extends Rule[LogicalPlan] with PredicateHelper {
+
+  def apply(plan: LogicalPlan): LogicalPlan = plan transform {
+    case filter @ Filter(condition,
+    aggregate @ Aggregate(groupingExpressions, aggregateExpressions, grandChild)) =>
```
Comment (marmbrus) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala]: nit: indentation.  I'm not sure we have a strict rule, but no indent is kinda hard to follow.  I'd probably try to make it a tree if it fits?

```scala
case filter @ Filter(condition,
     aggregate @ Aggregate(groupingExpressions, aggregateExpressions, grandChild)) =>
```

or just 4 space indent?

**Comment by marmbrus** [comment]
This looks great, thanks for doing it!  

Can you cleanup the title `[SPARK-11179] [SQL] Push filters through aggregate`
and the description: `Push conjunctive predicates though Aggregate operators when their references are a subset of the groupingExpressions.`

If you are feeling ambitious I'd include query plans before and after the optimization too.

These become the commit message when we use our merge tool.

**Comment by nitin2goyal** [comment]
Addressed 1 code review comment, fixed couple of scalastyle issues and cleaned up title and description in latest commit and this PR.

**Comment by nitin2goyal** [comment]
@marmbrus Test failed again due to whitespace issue. I have already removed whitespace from that line and I am not getting scalastyle issue when compiling locally (don't see any whitespace in code review also). Am I missing something?


[error] /home/jenkins/workspace/SparkPullRequestBuilder/sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/FilterPushdownSuite.scala:664:0: Whitespace at end of line

**Comment by marmbrus** [comment]
try `sbt test:scalastyle`

**Comment by rxin** [comment]
You can run local style tests by

dev/lint-scala

**Comment by nitin2goyal** [comment]
Somehow, I wasn't getting scalastyle issue on my branch with above 2 commands. Cloned apache spark master, cherry-picked my changes and then I got the error. Checked-in the whitespace removal. Please test this.

**Comment by marmbrus** [comment]
Thanks, merging to master!
